src/traversal.rs

```diff
@@ -0,0 +1,134 @@
+    let root_key = path_split.get(0).ok_or(anyhow!("Invalid query path!"))?;
+    let table = doc.as_table_mut();
+    let item = table
+        .get_mut(root_key)
+        .ok_or(anyhow!("Missing table for traversal"))?;
+    do_traverse(&path_slice[1..], &mut At::Item::<'a>(item), op)
+}
+
+fn do_traverse(path: &[&str], item: &mut At, op: TraverseOps) -> Result<Json> {
+    if !path.is_empty() {
+        return item.fold(path[0], &path[1..], op);
+    }
+
+    match op {
+        TraverseOps::Get(callback) => callback(item),
+    }
+}
+
+impl At<'_> {
+    fn fold(&mut self, key: &str, path: &[&str], op: TraverseOps) -> Result<Json> {
+        match self {
+            At::Array(arr) => {
+                let index = key
+                    .parse::<usize>()
+                    .map_err(|_| anyhow!("Key is not a valid usize"))?;
+                let member = arr.get_mut(index).ok_or(anyhow!("Array out of range"))?;
+                do_traverse(path, &mut At::Value(member), op)
+            }
+            At::ArrayOfTables(aar) => {
+                let index = key
+                    .parse::<usize>()
+                    .map_err(|_| anyhow!("Key is not a valid usize"))?;
+                let member = aar.get_mut(index).ok_or(anyhow!("Array out of range"))?;
+                do_traverse(path, &mut At::Table(member), op)
+            }
+            At::Item(item) => match item {
+                Item::ArrayOfTables(aar) => At::ArrayOfTables(aar).fold(key, path, op),
+                Item::Table(table) => At::Table(table).fold(key, path, op),
+                Item::Value(value) => At::Value(value).fold(key, path, op),
+                _ => Err(anyhow!("Unable to index item {:?} with {:?}", item, key)),
+            },
+            At::Table(table) => {
+                let mut found = table
+                    .get_mut(key)
+                    .ok_or(anyhow!("Unable to index table with {:?}", key))?;
+                do_traverse(path, &mut At::Item(&mut found), op)
+            }
+            At::Value(value) => match value {
+                Value::Array(arr) => At::Array(arr).fold(key, path, op),
+                _ => Err(anyhow!("Unable to index value {:?} with {:?}", value, key)),
+            },
+        }
+    }
+
+    pub fn to_value(&mut self) -> Result<serde_json::Value> {
+        match self {
+            At::Array(_arr) => Err(anyhow!("array")),
+            At::ArrayOfTables(_aar) => Err(anyhow!("stub")),
+            At::Item(item) => match item {
+                Item::None => Ok(Json::Null),
+                Item::Value(value) => At::Value(value).to_value(),
+                Item::ArrayOfTables(aar) => At::ArrayOfTables(aar).to_value(),
+                Item::Table(table) => At::Table(table).to_value(),
+            },
+            At::Value(value) => match value {
+                Value::String(s) => {
+                    s.fmt();
+                    Ok(Json::String(s.value().clone()))
+                }
```

Comment:
idt you have to `fmt` the string, but you'll want to do it on items
```suggestion
                Value::String(s) => Ok(Json::String(s.value().clone()),
```

Reply:
I was experiencing some oddities around `fmt()` where I was getting strings that incorrectly included formatting whitespace from the toml. I think I'll throw `fmt()` everywhere just in case.

Also write some tests.